runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -114,7 +144,528 @@ native2InterpJavaUpcallD(J9UpcallMetaData *data, void *argsListPointer)
+	J9Method* thrLiterals = NULL;
+	UDATA returnStorage = 0;
+	j9object_t sessionOrScopeObject = NULL;
+
+	/* Determine whether to use the current thread or create a new one
+	 * when there is no java thread attached to the native thread
+	 * created directly in native.
+	 */
+	currentThread = getCurrentThread(data, &isCurThrdAllocated);
+	if (NULL == currentThread) {
+		/* The OOM exception set in getCurrentThread() will be thrown in the interpreter
+		 * after returning from the native function in downcall.
+		 */
+		goto doneAndExit;
+	}
+
+	returnType = getReturnTypeFromMetaData(data);
+	returnsObject = ((J9NtcPointer == returnType) || (J9NtcStruct == returnType)) ? true : false;
+
+	VM_VMAccess::inlineEnterVMFromJNI(currentThread);
+	if (buildCallInStackFrameHelper(currentThread, &newELS, returnsObject)) {
+
+		/* Save the current executing method and restore later after the memory allocation
+		 * for pointer/struct in the argument list.
+		 */
+		thrLiterals = currentThread->literals;
+
+		/* The argument list of the upcall method handle on the stack includes the target method handle,
+		 * the method arguments and the appendix which is set via MethodHandleResolver.upcallLinkCallerMethod().
+		 *
+		 * Note: push the target method handle on the special frame so as to avoid updating the address
+		 * on the stack by GC when allocating memory for the next pointer/struct of the argument list.
+		 */
+		j9object_t calleeHandle = J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_CALLEEMH(currentThread, J9_JNI_UNWRAP_REFERENCE(data->mhMetaData));
+		PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, calleeHandle);
+
+		for (I_32 argIndex = 0; argIndex < paramCount; argIndex++) {
+			U_8 argSigType = sigArray[argIndex].type & J9_FFI_UPCALL_SIG_TYPE_MASK;
+
+			switch (argSigType) {
+			/* Small native types are saved at their natural boundaries (1, 2, 4-byte) in the stack on macOS/AArch64
+			 * while the stack slots are always 8-byte wide on Linux/AArch64.
+			 */
+#if defined(OSX) && defined(AARCH64)
+			case J9_FFI_UPCALL_SIG_TYPE_CHAR:
+			{
+				I_8 argValue = *(I_8*)vmFuncs->getArgPointer(nativeSig, argsListPointer, argIndex);
+				*(I_32*)--currentThread->sp = (I_32)argValue;
+				break;
+			}
+			case J9_FFI_UPCALL_SIG_TYPE_SHORT:
+			{
+				I_16 argValue = *(I_16*)vmFuncs->getArgPointer(nativeSig, argsListPointer, argIndex);
+				*(I_32*)--currentThread->sp = (I_32)argValue;
+				break;
+			}
+#else
+			case J9_FFI_UPCALL_SIG_TYPE_CHAR:  /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_SHORT: /* Fall through */
+#endif /* defined(OSX) && defined(AARCH64) */
+			case J9_FFI_UPCALL_SIG_TYPE_INT32: /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+			{
+				/* Convert the argument value to 64 bits prior to the 32-bit conversion to get the actual value
+				 * in the case of boolean/byte/char/short/int regardless of the endianness on platforms.
+				 */
+				I_64 argValue = *(I_64*)vmFuncs->getArgPointer(nativeSig, argsListPointer, argIndex);
```

Comment:
you dont need to go through the function table since you're in the vm dir

Reply:
Agreed and removed `vmFuncs->` at multiple places.